pilot/pkg/serviceregistry/kube/mcs/discoverability.go

```diff
@@ -0,0 +1,310 @@
+	monitoring.MustRegister(serviceExportEvents)
+}
+
+func incrementEvent(event string, cluster kube.ClusterName) {
+	serviceExportEvents.With(eventTag.Value(event), clusterTag.Value(string(cluster))).Increment()
+}
+
+type DiscoverabilityConfig struct {
+	Clusters     *kube.ClusterSet
+	XDSUpdater   model.XDSUpdater
+	DomainSuffix string
+}
+
+// NewDiscoverabilityController creates a new DiscoverabilityController that observes the given ClusterSet.
+func NewDiscoverabilityController(cfg DiscoverabilityConfig) *DiscoverabilityController {
+	c := &DiscoverabilityController{
+		clusters:     cfg.Clusters,
+		xdsUpdater:   cfg.XDSUpdater,
+		domainSuffix: cfg.DomainSuffix,
+		exports:      make(map[kube.ClusterName]*exportedServices),
+		queue:        queue.NewQueue(time.Second),
+	}
+
+	c.clusters.AddClusterLifecycleListener(func(e kube.ClusterEvent, name kube.ClusterName) {
+		switch e {
+		case kube.ClusterAdded:
+			c.onClusterAdded(name)
+		case kube.ClusterDeleted:
+			c.onClusterDeleted(name)
+		}
+	})
+	return c
+}
+
+// DiscoverabilityController is a ServiceDiscoverabilityPolicy that is driven by MCS resources in each cluster.
+type DiscoverabilityController struct {
+	clusters     *kube.ClusterSet
+	xdsUpdater   model.XDSUpdater
+	domainSuffix string
+	queue        queue.Instance
+	exports      map[kube.ClusterName]*exportedServices
+
+	// mutex for access to the exports map. Updates to cluster and ServiceExports
+	// should be infrequent enough that we can use a single lock for the entire map.
+	mutex sync.RWMutex
+}
+
+// Run this controller.
+func (c *DiscoverabilityController) Run(stopCh <-chan struct{}) error {
+	log.Debugf("%s started", discoverabilityLogPrefix)
+	go c.queue.Run(stopCh)
+	return nil
+}
+
+func (c *DiscoverabilityController) onClusterAdded(name kube.ClusterName) {
+	c.queue.Push(func() error {
+		log.Debugf("%s adding cluster %s", discoverabilityLogPrefix, name)
+
+		cluster := c.clusters.GetClusterWithName(name)
+		if cluster == nil {
+			log.Warnf("%s failed adding cluster %s: cluster not found", discoverabilityLogPrefix, name)
+			return nil
+		}
+
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+
+		// Create the informer and register callbacks.
+		informer := cluster.Client().MCSApisInformer().Multicluster().V1alpha1().ServiceExports().Informer()
+		informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+			AddFunc: func(obj interface{}) {
+				c.onServiceExportAdded(name, obj)
+			},
+			DeleteFunc: func(obj interface{}) {
+				c.onServiceExportDeleted(name, obj)
+			},
+		})
+
+		// Create a new map entry for this cluster.
+		c.exports[name] = newExportedServices(informer)
+		return nil
+	})
+}
+
+func (c *DiscoverabilityController) onClusterDeleted(name kube.ClusterName) {
+	c.queue.Push(func() error {
+		log.Debugf("%s deleting cluster %s", discoverabilityLogPrefix, name)
+		// Lock for write.
+		c.mutex.Lock()
+
+		// Delete the map entry for this cluster.
+		delete(c.exports, name)
+		c.mutex.Unlock()
```

Comment:
nit: might be more consistent to do this in a defer

Reply:
I prefer using defer when there are multiple return paths. There is a slight performance impact with using defer ... it likely wouldn't be an issue here, but I do use it that way fairly consistently.